jax/_src/extend/ffi.py

```diff
@@ -15,11 +15,103 @@
 from __future__ import annotations
 
 import os
+from collections.abc import Iterable, Mapping, Sequence
+from typing import Any
 
+import numpy as np
+
+from jax._src import core
+from jax._src import dtypes
+from jax._src.interpreters import mlir
 from jax._src.lib import jaxlib
+from jax._src.lib.mlir import ir
+from jax._src.typing import DimSize
 
 
 def include_dir() -> str:
   """Get the path to the directory containing header files bundled with jaxlib"""
   jaxlib_dir = os.path.dirname(os.path.abspath(jaxlib.__file__))
   return os.path.join(jaxlib_dir, "include")
+
+
+def ffi_lowering(
+    call_target_name: str,
+    *,
+    operand_layouts: Sequence[Sequence[DimSize]] | None = None,
+    result_layouts: Sequence[Sequence[DimSize]] | None = None,
+    backend_config: Mapping[str, ir.Attribute] | None = None,
+    **lowering_args: Any
+) -> mlir.LoweringRule:
+  """Build a lowering rule for an foreign function interface (FFI) target
+
+  By default, this lowering rule can use the input and output abstract values to
+  compute the input and output types and shapes for the custom call, assuming
+  row-major layouts.
+
+  If keyword arguments are passed at bind time, these are treated as attributes,
```

Comment:
I would guess that "bind time" might not be clear to users. One way to avoid it is to talk about a primitive and its parameters.

We defined primitive [https://jax.readthedocs.io/en/latest/glossary.html#term-primitive] here, but unfortunately don't mention parameters atm.